### PR TITLE
[FSDP2] Added pre/post-all-gather extensions (subclass)

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120256
* #123953
* __->__ #122908
* #119302

**Overview**
This PR adds pre/post-all-gather extensions to FSDP2.
- The pre/post-all-gather extensions are specified at the tensor-level on the `sharded_param._local_tensor` (i.e. the tensor wrapped by the sharded `DTensor`). If the user has a tensor-subclass parameter on the module passed to FSDP that preserves the subclass through the sharding ops (e.g. `new_zeros`, `chunk`, etc.), then the `sharded_param._local_tensor` will naturally be of that subclass.
- The pre-all-gather function has signature:
  ```
  def fsdp_pre_all_gather(self) -> Tuple[Tuple[torch.Tensor, ...], Any]
  ```
    - The first return value is a `Tuple[torch.Tensor, ...]` of the all-gather inputs. It is a tuple since a subclass could contribute >1 inner tensors.
    - The second return value is any optional metadata needed to pass through to the post-all-gather.
- The post all-gather function has signature:
  ```
  def fsdp_post_all_gather(
      self,
      all_gather_outputs: Tuple[torch.Tensor, ...],
      metadata: Any,
      param_dtype: torch.dtype,
      *,
      out: Optional[torch.Tensor] = None,
  ) -> Union[Tuple[torch.Tensor, Tuple[torch.Tensor, ...]], None]:
  ```
    - The `all_gather_outputs` are exactly the all-gathered versions of the `fsdp_pre_all_gather` 1st return value (representing the all-gather inputs). We make sure to unflatten these back to ND for the user.
    - The `metadata` is the `fsdp_pre_all_gather` 2nd return value, untouched.
    - The `param_dtype` is the parameter dtype based on the passed-in `MixedPrecisionPolicy`. Namely, if no policy is passed in, then `param_dtype` is the original dtype, and otherwise, it is the `MixedPrecisionPolicy.param_dtype`.
    - If `out` is not specified, then the return value has type `Tuple[torch.Tensor, Tuple[torch.Tensor, ...]]`. The first tuple item is the unsharded parameter (e.g. re-wrapping into some subclass). The second tuple item is a tuple of unsharded inner tensors that FSDP should free during reshard. These should be derived from the all-gather outputs.
    - The `out` argument is required due to FSDP's `resize_` usage. We require an in-place variant for the backward all-gather. Here, `out` will be exactly the object returned as the first tuple item in the out-of-place variant mentioned before. The unsharded inner tensors will be allocated before calling `fsdp_post_all_gather`. When `out` is specified, the `fsdp_post_all_gather` should return `None`. If the post-all-gather does not do any out-of-place ops, then the `out` variant can just be a no-op since the unsharded inner tensors will be the same as the all-gather outputs, which FSDP directly writes to after all-gather. (E.g., this is the case for both float8 and `NF4Tensor`.)
- We check for `fsdp_pre_all_gather` and `fsdp_post_all_gather` directly via `hasattr` to accommodate monkey patching so that we do not strictly require the user to use a tensor subclass. The monkey patch must happen after the local tensors have been finalized (after applying FSDP and after any meta-device init).
- For now, we require that all gradients in one FSDP parameter group share the same dtype. This is fine for float8 and `NF4Tensor` use cases. If this requirement is too strict, then in the future we can issue 1 reduce-scatter per dtype per group.

**Design Notes**
- We assume that the `sharded_param._local_tensor` is padded on dim-0.
    - This assumption should not block immediate use cases, and when we pad the `DTensor._local_tensor` by default, this assumption will always be true.
    - This assumption allows us to call `sharded_param._local_tensor.fsdp_pre_all_gather()`; i.e. it tells us from which tensor object to invoke `fsdp_pre_all_gather()`.
    - Suppose we want to compose with CPU offloading. Then, CPU offloading's H2D copy should run first, i.e. `sharded_param._local_tensor.to("cuda").fsdp_pre_all_gather()`, where `_local_tensor.to("cuda")` should return an instance of the subclass so that it still defines `fsdp_pre_all_gather()`. Note that in this case, the subclass instance on GPU is a temporary, which means caching values on it would not be possible. One possibility would be to have `.to("cuda")` move any cached values too.
- `fsdp_post_all_gather` can either return an unsharded parameter that aliases with the all-gather output or does not alias, but there is no way to know a priori.
    - If the unsharded parameter aliases with the all-gather output, then we should _not_ free the all-gather output in `unshard`.
    - If the unsharded parameter does not alias with the all-gather output, then we prefer to free the all-gather output in `unshard` to avoid holding the unneeded temporary.
    - One approach is for eager-mode to check for this alias (by comparing data pointers). However, this might be adversarial to full-graph compilation. The compromise for simplicity can be to always free the all-gather output in `reshard`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -236,6 +242,24 @@ def init_dtype_attrs(self, mp_policy: MixedPrecisionPolicy):
         self.reduce_dtype = reduce_dtype
         # None indicates that the mixed precision is not enabled
 
+    def _init_extensions(self) -> None:
+        inner_tensor = self._sharded_local_tensor
+        has_fsdp_pre_all_gather = hasattr(inner_tensor, "fsdp_pre_all_gather")
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:247]: Note: I am thinking to always check this via `hasattr(inner_tensor, "fsdp_pre_all_gather")` to leave a backdoor if somehow we can monkey patch this method onto the tensor without making it a subclass.

File: test/distributed/_composable/fsdp/test_fully_shard_extensions.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    def world_size(self) -> int:
+        return 2
+
+    @property
+    def device(self) -> torch.device:
+        return torch.device("cuda:0")
+
+    @contextlib.contextmanager
+    def patch_two_tensor_fsdp_all_gather(self):
+        lock = threading.Lock()
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_extensions.py:60]: cc: @weifengpy Do we have any multi-threaded patching utilities? / Does this pattern look like something we already have a function for?

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 3]
```diff
@@ -270,6 +270,7 @@ def reshard(self):
                 return
         self._to_sharded()
 
+    @torch.compiler.disable()
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:273]: To-do: separate this out and do not include in this PR
(however, we need this to compile with float8)

File: test/distributed/_composable/fsdp/test_fully_shard_extensions.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+from torch.testing._internal.two_tensor import TwoTensor
+
+
+def two_tensor_fsdp_pre_all_gather(self) -> Tuple[Tuple[torch.Tensor, ...], Any]:
+    all_gather_inputs = (self.a, self.b)
+    metadata = None
+    return all_gather_inputs, metadata
+
+
+def two_tensor_fsdp_post_all_gather(
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_extensions.py:31]: This `TwoTensor` example covers two cases:
1. A wrapper subclass that has >1 inner tensor: FSDP can still reconstruct this correctly!
2. A post-all-gather that is out-of-place (when `param_dtype` is different from the original dtype): FSDP should manage these different tensors correctly!

As mentioned in the PR description, for simplicity, `all_gather_outputs` will not be freed until the reshard, even if they do not alias the `unsharded_parameter` (like in this case with `param_dtype=torch.bfloat16`).

I did not test that memory behavior here since it is kind of involved to write such a unit test. However, I do have a unit test in the float8_experimental side to make sure that the memory usage is as expected.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 5]
```diff
@@ -236,6 +254,24 @@ def init_dtype_attrs(self, mp_policy: MixedPrecisionPolicy):
...
+        has_fsdp_post_all_gather = hasattr(inner_tensor, "fsdp_post_all_gather")
+        if has_fsdp_pre_all_gather != has_fsdp_post_all_gather:
+            raise AssertionError(
+                "Both fsdp_pre_all_gather and fsdp_post_all_gather should be defined "
+                f"if using all-gather extensions: {inner_tensor}"
+            )
+        if has_fsdp_pre_all_gather:
+            if self.padded_sharded_param_size != self._sharded_local_tensor.size():
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:267]: We make this assumption for now until `DTensor` always pads its local tensor. In particular, we want to run `fsdp_pre_all_gather()` from the ND sharded parameter.

**Comment by awgu** [comment]
Tagging @yf225 @bdhirsh in case this change has any impact on tracing

I changed the implementation from the original one so that we do not check for aliasing between `all_gather_outputs` and `unsharded_param`'s inner tensors since I thought that that would most certainly break tracing. The compromise is that all-gather outputs are always freed in reshard even if they do not alias with `unsharded_param` (so memory could be held onto a bit longer than needed).

File: test/distributed/_composable/fsdp/test_fully_shard_extensions.py (hunk 2 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_extensions.py:70]: nice! this is a good addon. I do not have similar utilities since I only read and write existing fields instead of deleting

**Comment by weifengpy** [comment]
thanks for making this happen so quickly! I will be testing this with NF4Tensor this week

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 3 above)
Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:273]: will be landing ``@torch.compiler.disable_if_config_true("skip_fsdp_hooks")`` this week https://github.com/pytorch/pytorch/pull/123021

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 6]
```diff
@@ -140,4 +140,9 @@ def _move_states_to_device(
             raise AssertionError(
                 f"Expects DTensor to be moved to {dtensor_mesh_type} but got {tensor.device}"
             )
-        tensor.data = tensor.to(device)
+        if is_traceable_wrapper_subclass(tensor):
+            with torch.no_grad():  # avoid autograd increasing C++ refcount by 1
+                tensor_on_device = nn.Parameter(tensor.to(device))
+            torch.utils.swap_tensors(tensor, tensor_on_device)
+        else:
+            tensor.data = tensor.to(device)
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:148]: just to reduce risk

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 7]
```diff
@@ -368,18 +439,33 @@ def to_sharded_post_forward_dtensor(self, tensor: torch.Tensor) -> DTensor:
...
 
     @property
     def all_gather_inputs(self) -> List[torch.Tensor]:  # 1D
         self._assert_in_states(ShardedState.SHARDED, ShardedState.SHARDED_POST_FORWARD)
         if self.sharded_state == ShardedState.SHARDED:
+            if hasattr(self._sharded_local_tensor, "fsdp_pre_all_gather"):
+                (
+                    all_gather_inputs,
+                    self._extensions_data.all_gather_metadata,
+                ) = self._sharded_local_tensor.fsdp_pre_all_gather()
```
[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:459]: will dig more into the difference to see should we change extension core or NF4Tensor

[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:459]: I think we want to call the extension on the ND tensor, not the 1D tensor, ideally, since some user pre-all-gather may depend on the ND shape.

[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:459]: it running good know. I am guessing world size now base on metadata (before all-gather) and inner tensor shapes (after all-gather). world_size is used to restore size(0) in outer wrapper class. wonder if we should make world_size accessible from the extension ?

**Review by weifengpy — APPROVED** [review]
we discussed 2 things
* pass fsdp device mesh to extention so we can use mesh size to restore dim-0 shape for outer wrapper class. leave to Andrew to decide if it should be a follow up PR
* ``@torch.compiler.disable()`` is likely not needed in this PR but is needed in extention for fp8

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 8]
```diff
@@ -411,25 +497,26 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
...
+    size = tensor.numel() * tensor.itemsize
+    if (storage := tensor.untyped_storage()).size() != size:
+        storage.resize_(size)
 
 
-def unsafe_free_storage(tensor: torch.Tensor) -> None:
-    # Skip the already-freed check to save CPU overhead
-    tensor.untyped_storage().resize_(0)
+def free_storage(tensor: torch.Tensor) -> None:
+    if (storage := tensor.untyped_storage()).size() != 0:
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:518]: 🤔 is the bracket here correct? seems the left most one does not have a match

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 9]
```diff
@@ -93,6 +94,18 @@ class ParamModuleInfo:
     shared_param_names: List[str] = field(default_factory=list)
 
 
+@dataclass
+class ExtensionsData:
+    # User-defined metadata passed from pre to post-all-gather
+    all_gather_metadata: Optional[Any] = None
+    # Save the all-gather input sizes to unflatten the all-gather outputs to ND
+    all_gather_input_sizes: Sequence[torch.Size] = ()  # ND
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:102]: wondering this `ExtensionData` can be merged together with the default FSDPParam metas?

Looks like simple `torch.Tensor` case is just one case of this

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 10]
```diff
@@ -251,13 +287,40 @@ def init_all_gather_outputs(
         ]
 
     def init_unsharded_param(self):
-        if hasattr(self, "_unsharded_param"):
-            return  # already initialized
-        # For the default path (no post-all-gather), the all-gather output
-        # gives the unsharded parameter data directly
-        assert len(self.all_gather_outputs) == 1
+        if hasattr(self, "_unsharded_param"):  # after the 1st all-gather
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:290]: Might be similar to my previous comment, I feel this can be simplified by treating `torch.Tensor` and tensor subclass the same way, we can always check if the inner tensor have fsdp_pre_all_gather and fsdp_post_all_gather, if so we do additional handling, if not they are just vanilla tensor path?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 11]
```diff
@@ -222,13 +221,25 @@ def foreach_reduce_scatter_copy_in(
...
+    all_gather_dtype = param_all_gather_inputs[0][0].dtype
     for all_gather_inputs in param_all_gather_inputs:
-        param_all_gather_input_dtypes.append([t.dtype for t in all_gather_inputs])
-        param_all_gather_input_numels.append([t.numel() for t in all_gather_inputs])
-    return param_all_gather_input_dtypes, param_all_gather_input_numels
+        input_dtypes: List[torch.dtype] = []
+        input_numels: List[int] = []
+        for all_gather_input in all_gather_inputs:
+            if all_gather_input.dtype != all_gather_dtype:
+                all_gather_dtype = torch.uint8
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:233]: so this is handling the case where a FSDPParamGroup consists of different tensor dtypes, where when we perform allgather on this group, we just treat them universally as uint8 and recover their dtypes later?

**Review by wanchaol — APPROVED** [review]
logically lgtm! have a few comments to see if we can further simplify the code

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:518]: It matches with the bracket right before `.size()` :) -- it is kind of confusing :/

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 11 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:233]: yep!

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 10 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:290]: let me think through this more and see if we can unify

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:102]: I think one thing is that the `ExtensionsData` will never be used for the simple `torch.Tensor` case:
- The `all_gather_metadata` is specific to the extensions so that the user can pass metadata between the pre and post-all-gather.
- The `all_gather_input_sizes` are not needed since we do not need to unflatten the all-gather outputs to ND (but rather just keep as 1D).

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 10 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:290]: I think that the code below is as unified as it can be 🤔 . The tail logic for constructing the unsharded parameter is shared.

The divergent part is required since for the extensions path we (1) unflatten all-gather outputs and (2) have to call `fsdp_post_all_gather` either with its out variant or not. There is no reason to unflatten all-gather outputs for the default `torch.Tensor` path, and it would be an extra view op.
